Flexible grammar installer

Mandarancio opened the pull request:
Instead of hard-coding the grammars and queries I added the possibilities to install them and add them in the `init.lua` user configuration file.

Now it is possible to configure its own grammar from local files or from remote git repositories (doc and examples in the `README.md`)

TorchedSammy reviewed (commented):
ok so i did a quick check of this PR, and i have a few things to say.

firstly, i've been planning to add this feature *eventually* so thanks for deciding to work on it! the problem is that this removes "officially" supported queries and parsers. queries are taken from nvim-treesitter and are more complete than those hosted on the repository of a parser. if you could add this feature without removing officially supported parsers, that would be cool.

secondly, why did you remove the installer? the reason it downloads from the build repository is that windows users can't easily just compile a parser.

thirdly, functions should be camelCase and commits should be [conventional](https://www.conventionalcommits.org/en/v1.0.0/).

TorchedSammy commented:
you could've used the existing installer functions to facilitate this feature, no?

i've also realized that this removes defaults from evergreen and users need to find the parsers to use for their languages. it adds extra effort for no good reason.

i would prefer to not remove the ease of use and only add the ability to configure extra parsers (which would be none in the future, optimally) and not all required ones of a user.

Mandarancio commented:
Hi @TorchedSammy, thank you for your feedback!
I did not think about Windows users (as I am not used to deal with them ;) ) but your concern makes completely sense!
I have the following proposal:
 1.  Camel case function naming does not seem to be the standard as [`lite-xl-lsp`](https://github.com/lite-xl/lite-xl-lsp/) or [`console`](https://github.com/lite-xl/console/) use snake case, and these are officially supported plugin (or at least managed by the core team). So I believe this should be the convention (or at least for the functions that are called in the `init.lua`). For the commits I never heard about conventional and I will look into it. Maybe it can be useful to have some sort of guidelines about it in the project.
 2.  I think I could have a solution on how to merge the simplicity and flexibility: 
   1. I will add another installation method that will download the pre-compiled library
   2. I will add a structure (like yours) with some default languages and their configuration if the user adds a language without specifying any path/url/git the plugin will try to use the default config (if any)
   3. I will add back the default queries as for the previous point if the user does not specify any queries parameter and an existing default query exists this one will be used 
   4. I will add a command to display the status of the plugin (installed grammars, file extension and so on)
 3. I do not think having a command `install` is a good idea as it hides too much information to the user (as it is finally a hidden configuration) and it makes more difficult to manage the grammars I think that add the default and simply add `languages.add_language("c")` to the `init.lua` is easy enough as well as very clear.
 
Let me know what do you think about my proposal. 
Cheers!
Martino

TorchedSammy commented:
1. camel case is what i prefer for my code, so it will be what is used here.
2. all the suggestions here sound good, but let's see how it'll turn out?
3. i would prefer to keep the install command, no real reason to remove it. though other contributors can comment @xcb-xwii @takase1121

overall :+1:

Mandarancio commented:
I put back the queries and added the defaults installer back (look at the `README.md` for more information) and added the `status` command to check installed plugins, it shows the result (installed grammars and grammars that have been configured but failed to install) in the logs but I am not sure if it is the best solution.

I did not add back the install command as it will not fit the logic to manage the configurations (as the installation is made automatically at initialization if the library or the query is not found in the data folder). 
It could be add in the form of a work around (at initialization you could check for all the defaults languages and check if they are installed in the data folder) but is not very elegant.

I tried to use the `conventional` standard for the commits.

xcb-xwii commented:
> 3. i would prefer to keep the install command, no real reason to remove it. though other contributors can comment @xcb-xwii @takase1121

I also do not really see a reason to do so, but I do not mind removing it.

It might be useful to be able to install parsers on the fly without having to restart / reload, but it's not too big of a benefit to me.

However, if we introduce ways to update parsers, it might be worth it to have commands to manage parsers, so we can selectively update parsers, and also not have them all update immediately on startup.

---

There are some problems with the implementation here. This PR, as far as I can tell, does not attempt to check for updates / changes to the Git repo, and for the URL, it just re-downloads every time (do correct me if I'm wrong on these). Additionally, if a parser is already installed in a directory, replacing the parser with a different one inside the same directory requires you to remove the directory first, which is something we do not provide.

Another major issue right now is that we provide no way to update and remove parsers, which I've mentioned above. The more declarative installation methods implemented by this PR are nice, but it needs improvement.

xcb-xwii commented:
These changes seems to have removed the default Git repositories for the parsers. I think that needs to be restored, since not everyone wants to, or can use our prebuilt binaries, since only binaries for amd64 and select OSes are provided. A config option should also be added to choose between the prebuilt binaries or download from default repos for the built-in parsers.

Mandarancio commented on init.lua:
```diff
@@ -255,3 +261,35 @@ command.add('core.docview!', {
 		end
 	end
 })
+
+command.add(nil, {
+	['evergreen:status'] = function()
+		local notInstalled = {}
+		local installed = {}
+		local errorCounter = 0
+		for lang, options in pairs(languages.grammars) do
+			local lib = util.join { config.parserLocation, lang, "parser.so" }
+			local queries = util.join { config.queryLocation, lang, "highlights.scm" }
+			if not util.exists(lib) or not util.exists(queries) then
+				errorCounter = errorCounter + 1
+				table.insert(notInstalled, lang)
+			else
+				local str = " - " .. lang .. ": "
+				if options.extensions ~= nil then
+					str = str .. options.extensions .. " "
+				end
+				if options.filename ~= nil then
+					str = str .. options.filename .. " "
+				end
+				if options.filename == nil and options.extensions == nil then
+					str = str .. lang
+				end
+				table.insert(installed, str)
+			end
+		end
+		core.log("[Evergreen] Installed grammars:\n%s", table.concat(installed, "\n"))
+		if errorCounter > 0 then
+			core.warn("[Evergreen] grammars not operational:\n%s", table.concat(notInstalled, "\n"))
+		end
+	end
+})
```
no problem, I will put single quotes

Mandarancio commented:
@xcb-xwii I do agree with you that something to manage parsers (at least to updates and clean the ones no more used) should be implemented (and I was planning to do so).
For the moment at the initialization I just check if for the languages configured exists the parser library and the queries, if not I try to install them using the user configuration. But this is somehow similar to how it is done now (in the sense that evergeen does not check if there is a new version of the installed parsers or anything else, but it just installs the current version whenever the installation command is issued). 
Check automatically for updates could slow down the initialization of lite (in particular if many parsers are configured), so I propose to add an `update` and a `clean` command to do so.

Mandarancio commented:
> These changes seems to have removed the default Git repositories for the parsers. I think that needs to be restored, since not everyone wants to, or can use our prebuilt binaries, since only binaries for amd64 and select OSes are provided. A config option should also be added to choose between the prebuilt binaries or download from default repos for the built-in parsers.

Good suggestion

Mandarancio commented on LICENSE:
```diff
@@ -19,3 +19,4 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
+
```
I do not think it is really an issue, but I will revert it.

Mandarancio commented:
> > These changes seems to have removed the default Git repositories for the parsers. I think that needs to be restored, since not everyone wants to, or can use our prebuilt binaries, since only binaries for amd64 and select OSes are provided. A config option should also be added to choose between the prebuilt binaries or download from default repos for the built-in parsers.
> 
> Good suggestion

@xcb-xwii now there is the possibility to chose to use the precompiled parser or the default git for the "supported" languages (by default, if not specify it will use the precompiled)

Mandarancio commented:
> @xcb-xwii I do agree with you that something to manage parsers (at least to updates and clean the ones no more used) should be implemented (and I was planning to do so). For the moment at the initialization I just check if for the languages configured exists the parser library and the queries, if not I try to install them using the user configuration. But this is somehow similar to how it is done now (in the sense that evergeen does not check if there is a new version of the installed parsers or anything else, but it just installs the current version whenever the installation command is issued). Check automatically for updates could slow down the initialization of lite (in particular if many parsers are configured), so I propose to add an `update` and a `clean` command to do so.

@xcb-xwii, @TorchedSammy I added the following commands:
 - `evergreen:update` to update a single parser/queries
 - `evergreen:update-all` to update all parsers/queries
 - `evergreen:clean` to clean up unused parsers/queries
 - `evergreen:status` to check the available grammars

I do not think to implement any other features, as I think is now quite complete and flexible. 
Best luck!

TorchedSammy commented on installer.lua:
```diff
@@ -1,101 +1,275 @@
+    git = 'https://github.com/tree-sitter/tree-sitter-cpp'
+  },
+  d = {
+    precompiled = true,
```
d is no longer precompiled (at least for now)

TorchedSammy commented on installer.lua:
```diff
@@ -1,101 +1,275 @@
+  local tmp_path = util.join { config.dataDir, 'temp' }
+  system.mkdir(tmp_path)
+  local repo_path = util.join { tmp_path, options.lang }
+  if util.isDir(repo_path) then
+    core.log('[Evergreen] repository path "%s" exists', repo_path)
+    return true
+  end
+  core.log('[Evergreen] cloning ' .. options.git .. ' in ' .. repo_path)
+  util.exec({ 'git', 'clone', options.git, options.lang }, { cwd = tmp_path })
+  if options.rev ~= nil then
+    core.log('[Evergreen] checkout revision ' .. options.rev)
+    util.exec({ 'git', 'checkout', options.rev }, { cwd = repo_path })
+  end
```
these logs would be more suited as debug (so make a silent log)

TorchedSammy commented on util.lua:
```diff
@@ -14,4 +14,39 @@ function M.join(parts)
+  return nil
+end
+
+-- remove directory
+function M.rmDir(path)
+  M.exec(PLATFORM == 'Windows' and
+    { 'cmd', '/c', 'rmdir ' .. path } or
+    { 'sh', '-c', 'rm -rf ' .. path })
+end
```
isn't there a function for this in lite xl core?

Mandarancio replied:
There is  the following function:
 - `system.rmdir(dir)` removes a directory. Use this instead of `os.remove().` but **The directory must be empty.**

As the directory will not be empty I wrote this stupid (and ugly) function...

takase1121 replied:
1. This is highly unsafe. You didn't even attempt to quote the parameters.
2. You don't need to start a shell for this. Just `rm` is sufficient.

Mandarancio replied:
I do agree, I did copy some snippet from somewhere to quickly develop a functional tool, I rewrote using lite-xl system functions

TorchedSammy commented on README.md:
```diff
@@ -14,21 +14,98 @@ It is work in progress, but functions well.
+It is possible to use grammar installed locally or directly from their git repository. Additionally 
+it is possible to specify where to find the queries of the language as well as use sub path for projects
+containing more then one language.
+## Default grammars
+
+A few grammars have default configurations, pre-compiled parsers and custom queries:
+
+The supported languages are:
+ 
```
this can be formatted a bit better:

```suggestion
## Default grammars

A few grammars have default configurations, pre-compiled parsers and custom queries.
The supported languages are:  
```

TorchedSammy commented on README.md:
````diff
@@ -14,21 +14,98 @@ It is work in progress, but functions well.
+
+To use any of this grammars simply add this to your conifg 
+
+```lua
+local egconfig = require "plugins.evergreen.config"
+
+egconfig.addGrammar {
+  lang = "LANGUAGE",
+  precompiled = true -- optional: default true, if false use default git repository to install grammar
+}
+```
````
change explanation of precompiled option as seen fit, but this is what i would write

```suggestion
```lua
local config = require "plugins.evergreen.config"

config.addGrammar {
  lang = "LANGUAGE",
  precompiled = true -- defaults to true, specifies if language parser is precompiled and available on the evergreen builds repo. all supported languages have precompiled parsers.
}
```
```

TorchedSammy commented on README.md:
````diff
@@ -14,21 +14,98 @@ It is work in progress, but functions well.
+
+## From git repository 
+To add a grammar directly from a git repository add this to your `init.lua` scripts:
+
+```lua
+local egconfig = require "plugins.evergreen.config"
+egconfig.addGrammar {
+  git = "GIT REPO",
+  lang = "LANGUAGE NAME",
+  extensions = "FILE EXTENSIONS", -- optional for extension name based
+  filename = "FILE NAME", -- optional for file name based 
+  queries = "QUERIES SUB PATH", -- optional sub path of grammar queries, default = 'queries'
+  subpath = "GRAMMAR SUB PATH", -- optional sub path if grammar is nested, default = nil
+  revision = "SPECIFIC GIT REVISION", -- optional version of the code to clone
+}
+```
+
+## Example 
+
+```lua
+local egconfig = require "plugins.evergreen.config"
+
+egconfig.addGrammar {
+  path = "~/.grammars/tree-sitter-epics/epics-db",
+  lang = "epics_db",
+  extensions = "db,vdb",
+}
+
+egconfig.addGrammar {
+  git = "https://github.com/tree-sitter/tree-sitter-cpp.git",
+  lang = "cpp",
+  extensions = "cpp,hpp,cc",
+}
+
+egconfig.addGrammar {
+  git = "https://github.com/tree-sitter/tree-sitter-c.git",
+  lang = "c",
+  extensions = "c,h",
+}
+
````
this entire section should be condensed so that the examples will also specify all available values.

Mandarancio replied:
Now I am using tables for both `extensions` and `filenames`. Now I am missing to rewrite the example section.

Mandarancio replied:
I simplify the readme and the example section I hope it is now better.

TorchedSammy reviewed (changes requested):
lost a bit of interest in evergreen for a bit. still haven't tried out the changes myself. mainly cosmetic stuff

TorchedSammy commented on README.md:
```diff
@@ -14,25 +14,78 @@ It is work in progress, but functions well.
+-- you can simply install it manually
+egconf.addGrammar {
+	lang = "epics_db",
+	path = "$HOME/.grammars/tree-sitter-epics/epics-db", -- path where it is installed 
```
does `~` not work in the path? that would be better.

Mandarancio replied:
I fixed the code and now `~` work (and actually before nor `$HOME` nor `~` were working)

TorchedSammy commented on README.md:
```diff
@@ -14,25 +14,78 @@ It is work in progress, but functions well.
+  extensions = { "cue" }, -- list of extensions to match.
+                          -- it can be omitted if the language has the same name of the
+                          -- extension (like in this case)
+  -- filenames = {...}, -- list of filenames to match
```
better explanation? what are the filenames matching for..

Mandarancio replied:
The file names are  coming from the current implementation of evergreen and in particular at `languages.lua:30`.
But basically is for grammars that matches complete filenames and not only extensions, but I believe it would be better to add regular expression (such as `%.cue`) in order to first be more flexible (for example I have makesfile that could be matched  by something like this `Makefile.%`) and avoid the need to have the two separate fields. Moreover I think `lite-xl-lsp` uses something like that.

What do you think about it?

Mandarancio replied:
Hi, I updated it with a single `filePatterns` entry where it is possible to specify any regex (as in `lite-xl-lsp:file_patterns`  (is really unfortunate to use different naming patterns...)). 
Please let me know what do you think about it.

TorchedSammy replied:
assuming this is the same as ignore_patters in lite xl, it's a good change.

TorchedSammy reviewed (commented):
that is all from me i think, @takase1121 @xcb-xwii review
